File: Collections/INotifyingCollection.vb
Imports System.Collections.Specialized

Namespace Common

    Public Delegate Sub NotifyCollectionChangedEventHandler(Of T)(sender As Object, e As NotifyCollectionChangedEventArgs(Of T))

    Public Class NotifyCollectionChangedEventArgs(Of T)
        Inherits NotifyCollectionChangedEventArgs

#Disable Warning BC40000 ' Type or member is obsolete
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction)
            MyBase.New(action)

            InitializeAdd(action, Nothing, -1)
        End Sub

        Public Shared Function CreateReset() As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Reset)
        End Function

        ''' <summary>
        ''' For Add, Remove, or Reset.
        ''' </summary>
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction, changedItem As T)
            MyBase.New(action, changedItem)

            If action = NotifyCollectionChangedAction.Reset Then
                InitializeAdd(action, Nothing, -1)
            Else
                InitializeAddOrRemove(action, New T() {changedItem}, -1)
            End If
        End Sub

        Public Shared Function CreateAdd(changedItem As T) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Add, changedItem)
        End Function

        Public Shared Function CreateRemove(changedItem As T) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Remove, changedItem)
        End Function

        ''' <summary>
        ''' For Add, Remove, or Reset.
        ''' </summary>
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction, changedItem As T, index As Integer)
            MyBase.New(action, changedItem, index)

            If action = NotifyCollectionChangedAction.Reset Then
                InitializeAdd(action, Nothing, -1)
            Else
                InitializeAddOrRemove(action, New T() {changedItem}, index)
            End If
        End Sub

        Public Shared Function CreateAdd(changedItem As T, index As Integer) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Add, changedItem, index)
        End Function

        Public Shared Function CreateRemove(changedItem As T, index As Integer) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Remove, changedItem, index)
        End Function

        ''' <summary>
        ''' For Add, Remove, or Reset.
        ''' </summary>
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction, changedItems As IList(Of T))
            MyBase.New(action, DirectCast(changedItems, IList))

            If action = NotifyCollectionChangedAction.Reset Then
                InitializeAdd(action, Nothing, -1)
            Else
                InitializeAddOrRemove(action, changedItems, -1)
            End If
        End Sub

        Public Shared Function CreateAdd(changedItems As IList(Of T)) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Add, changedItems)
        End Function

        Public Shared Function CreateRemove(changedItems As IList(Of T)) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Remove, changedItems)
        End Function

        ''' <summary>
        ''' For Add, Remove, or Reset.
        ''' </summary>
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction, changedItems As IList(Of T), startingIndex As Integer)
            MyBase.New(action, DirectCast(changedItems, IList), startingIndex)

            If action = NotifyCollectionChangedAction.Reset Then
                InitializeAdd(action, Nothing, -1)
            Else
                InitializeAddOrRemove(action, changedItems, startingIndex)
            End If
        End Sub

        Public Shared Function CreateAdd(changedItems As IList(Of T), startingIndex As Integer) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Add, changedItems, startingIndex)
        End Function

        Public Shared Function CreateRemove(changedItems As IList(Of T), startingIndex As Integer) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Remove, changedItems, startingIndex)
        End Function

        ''' <summary>
        ''' For Move or Replace.
        ''' </summary>
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction, newItem As T, oldItem As T)
            MyBase.New(action, newItem, oldItem)

            InitializeMoveOrReplace(action, New T() {newItem}, New T() {oldItem}, -1, -1)
        End Sub

        Public Shared Function CreateReplace(newItem As T, oldItem As T) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Replace, newItem, oldItem)
        End Function

        ''' <summary>
        ''' For Move or Replace.
        ''' </summary>
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction, newItem As T, oldItem As T, index As Integer)
            MyBase.New(action, newItem, oldItem, index)

            InitializeMoveOrReplace(action, New T() {newItem}, New T() {oldItem}, index, index)
        End Sub

        Public Shared Function CreateReplace(newItem As T, oldItem As T, index As Integer) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Replace, newItem, oldItem, index)
        End Function

        ''' <summary>
        ''' For Move or Replace.
        ''' </summary>
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction, newItems As IList(Of T), oldItems As IList(Of T))
            MyBase.New(action, DirectCast(newItems, IList), DirectCast(oldItems, IList))

            InitializeMoveOrReplace(action, newItems, oldItems, -1, -1)
        End Sub

        Public Shared Function CreateMove(changedItem As T) As NotifyCollectionChangedEventArgs(Of T)
            Dim Items = New T() {changedItem}
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Move, Items, Items)
        End Function

        Public Shared Function CreateMove(changedItems As IList(Of T)) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Move, changedItems, changedItems)
        End Function

        Public Shared Function CreateReplace(newItems As IList(Of T), oldItems As IList(Of T)) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Replace, newItems, oldItems)
        End Function

        ''' <summary>
        ''' For Move or Replace.
        ''' </summary>
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction, newItems As IList(Of T), oldItems As IList(Of T), startingIndex As Integer)
            MyBase.New(action, DirectCast(newItems, IList), DirectCast(oldItems, IList), startingIndex)

            InitializeMoveOrReplace(action, newItems, oldItems, startingIndex, startingIndex)
        End Sub

        Public Shared Function CreateReplace(newItems As IList(Of T), oldItems As IList(Of T), startingIndex As Integer) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Replace, newItems, oldItems, startingIndex)
        End Function

        ''' <summary>
        ''' For Move or Replace.
        ''' </summary>
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction, changedItem As T, index As Integer, oldIndex As Integer)
            MyBase.New(action, changedItem, index, oldIndex)

            Dim changedItems = New T() {changedItem}
            InitializeMoveOrReplace(action, changedItems, changedItems, index, oldIndex)
        End Sub

        Public Shared Function CreateMove(changedItem As T, index As Integer, oldIndex As Integer) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Move, changedItem, index, oldIndex)
        End Function

        ''' <summary>
        ''' For Move or Replace.
        ''' </summary>
        <Obsolete("Use the Create methods instead.")>
        Public Sub New(action As NotifyCollectionChangedAction, changedItems As IList(Of T), index As Integer, oldIndex As Integer)
            MyBase.New(action, DirectCast(changedItems, IList), index, oldIndex)

            InitializeMoveOrReplace(action, changedItems, changedItems, index, oldIndex)
        End Sub

        Public Shared Function CreateMove(changedItems As IList(Of T), index As Integer, oldIndex As Integer) As NotifyCollectionChangedEventArgs(Of T)
            Return New NotifyCollectionChangedEventArgs(Of T)(NotifyCollectionChangedAction.Move, changedItems, index, oldIndex)
        End Function
#Enable Warning BC40000 ' Type or member is obsolete

        Public Shared Function FromNotifyCollectionChangedEventArgs(ByVal E As NotifyCollectionChangedEventArgs) As NotifyCollectionChangedEventArgs(Of T)
            Select Case E.Action
                Case NotifyCollectionChangedAction.Move
                    Return CreateMove(E.NewItems.Cast(Of T)().ToArray(),
                                  E.NewStartingIndex,
                                  E.OldStartingIndex)
                Case NotifyCollectionChangedAction.Replace
                    Return CreateReplace(E.NewItems.Cast(Of T)().ToArray(),
                                     E.OldItems.Cast(Of T)().ToArray(),
                                     E.NewStartingIndex)
                Case NotifyCollectionChangedAction.Reset
                    Return CreateReset()
                Case NotifyCollectionChangedAction.Add
                    Return CreateAdd(E.NewItems.Cast(Of T)().ToArray(),
                                 E.NewStartingIndex)
                Case NotifyCollectionChangedAction.Remove
                    Return CreateRemove(E.OldItems.Cast(Of T)().ToArray(),
                                    E.OldStartingIndex)
            End Select

            Assert.Fail()
            Return Nothing
        End Function

        Public Shared Function FromNotifyCollectionChangedEventArgsNoIndex(ByVal E As NotifyCollectionChangedEventArgs) As NotifyCollectionChangedEventArgs(Of T)
            Select Case E.Action
                Case NotifyCollectionChangedAction.Move
                    Return CreateMove(E.NewItems.Cast(Of T)().ToArray())
                Case NotifyCollectionChangedAction.Replace
                    Return CreateReplace(E.NewItems.Cast(Of T)().ToArray(),
                                     E.OldItems.Cast(Of T)().ToArray())
                Case NotifyCollectionChangedAction.Reset
                    Return CreateReset()
                Case NotifyCollectionChangedAction.Add
                    Return CreateAdd(E.NewItems.Cast(Of T)().ToArray())
                Case NotifyCollectionChangedAction.Remove
                    Return CreateRemove(E.OldItems.Cast(Of T)().ToArray())
            End Select

            Assert.Fail()
            Return Nothing
        End Function

        Private Sub InitializeAddOrRemove(action As NotifyCollectionChangedAction, changedItems As IList(Of T), startingIndex As Integer)
            If action = NotifyCollectionChangedAction.Add Then
                InitializeAdd(action, changedItems, startingIndex)
            ElseIf action = NotifyCollectionChangedAction.Remove Then
                InitializeRemove(action, changedItems, startingIndex)
            End If
        End Sub

        Private Sub InitializeAdd(action As NotifyCollectionChangedAction, newItems As IList(Of T), newStartingIndex As Integer)
            _NewItems = If((newItems Is Nothing), Nothing, New List(Of T)(newItems).AsReadOnly())
        End Sub

        Private Sub InitializeRemove(action As NotifyCollectionChangedAction, oldItems As IList(Of T), oldStartingIndex As Integer)
            _OldItems = If((oldItems Is Nothing), Nothing, New List(Of T)(oldItems).AsReadOnly())
        End Sub

        Private Sub InitializeMoveOrReplace(action As NotifyCollectionChangedAction, newItems As IList(Of T), oldItems As IList(Of T), startingIndex As Integer, oldStartingIndex As Integer)
            InitializeAdd(action, newItems, startingIndex)
            InitializeRemove(action, oldItems, oldStartingIndex)
        End Sub

#Region "ItemsGotIn Property"
        Public ReadOnly Property ItemsGotIn As IList(Of T)
            Get
                If Me.Action = NotifyCollectionChangedAction.Add Or Action = NotifyCollectionChangedAction.Replace Then
                    Return Me.NewItems
                End If
                Return Utilities.Typed(Of T).EmptyArray
            End Get
        End Property
#End Region

#Region "ItemsWentOut Property"
        Public ReadOnly Property ItemsWentOut As IList(Of T)
            Get
                If Me.Action = NotifyCollectionChangedAction.Remove Or Action = NotifyCollectionChangedAction.Replace Then
                    Return Me.OldItems
                End If
                Return Utilities.Typed(Of T).EmptyArray
            End Get
        End Property
#End Region

#Region "NewItems Property"
        Private _NewItems As IList(Of T)

        Public Shadows ReadOnly Property NewItems As IList(Of T)
            Get
                Return Me._NewItems
            End Get
        End Property
#End Region

#Region "OldItems Property"
        Private _OldItems As IList(Of T)

        Public Shadows ReadOnly Property OldItems As IList(Of T)
            Get
                Return Me._OldItems
            End Get
        End Property
#End Region

    End Class

    Public Interface INotifyCollectionChanged(Of T)
        Inherits INotifyCollectionChanged

        Shadows Event CollectionChanged As NotifyCollectionChangedEventHandler(Of T)

    End Interface

    Public Interface INotifyingCollection(Of T)
        Inherits INotifyCollectionChanged(Of T),
             ICollection(Of T)

    End Interface

    Public Class NotifyingList(Of T)
        Inherits BaseList(Of T)
        Implements INotifyingCollection(Of T)

        Public Sub New(ByVal List As IList(Of T))
            Me.BaseList = List
        End Sub

        Public Sub New()
            Me.New(New List(Of T)())
        End Sub

        Public Overrides Sub Clear()
            Me.BaseList.Clear()
            Me.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateReset())
        End Sub

        Public Overrides Sub Insert(index As Integer, item As T)
            Me.BaseList.Insert(index, item)
            Me.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateAdd(item, index))
        End Sub

        Public Overrides Sub RemoveAt(index As Integer)
            Dim OldItem = Me.BaseList.Item(index)
            Me.BaseList.RemoveAt(index)
            Me.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateRemove(OldItem, index))
        End Sub

        Public Overridable Sub Move(ByVal OldIndex As Integer, NewIndex As Integer)
            Dim Item = Me.BaseList.Item(OldIndex)
            Me.BaseList.Move(OldIndex, NewIndex)
            Me.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateMove(Item, NewIndex, OldIndex))
        End Sub

        Public Overridable Sub SetFrom(ByVal Collection As IEnumerable(Of T))
            Me.BaseList.Clear()
            Me.BaseList.AddRange(Collection)
            Me.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateReset())
        End Sub

        Protected Overrides Function IEnumerable_1_GetEnumerator() As IEnumerator(Of T)
            Return Me.BaseList.GetEnumerator()
        End Function

        Public Function GetEnumerator() As IEnumerator(Of T)
            Return Me.BaseList.GetEnumerator()
        End Function

        Public Overrides ReadOnly Property Count As Integer
            Get
                Return Me.BaseList.Count
            End Get
        End Property

        Default Public Overrides Property Item(index As Integer) As T
            Get
                Return Me.BaseList.Item(index)
            End Get
            Set(value As T)
                Dim OldItem = Me.BaseList.Item(index)
                Me.BaseList.Item(index) = value
                Me.OnCollectionChanged(NotifyCollectionChangedEventArgs(Of T).CreateReplace(value, OldItem, index))
            End Set
        End Property

#Region "CollectionChanged Event"
        Public Shadows Event CollectionChanged As NotifyCollectionChangedEventHandler(Of T) Implements INotifyCollectionChanged(Of T).CollectionChanged
        Private Event INotifyCollectionChanged_CollectionChanged As NotifyCollectionChangedEventHandler Implements INotifyCollectionChanged.CollectionChanged

        Protected Overridable Sub OnCollectionChanged(ByVal E As NotifyCollectionChangedEventArgs(Of T))
            RaiseEvent INotifyCollectionChanged_CollectionChanged(Me, E)
            RaiseEvent CollectionChanged(Me, E)
        End Sub
#End Region

        Private ReadOnly BaseList As IList(Of T)

    End Class

End Namespace

File: Collections/INotifyingCollection.cs
using System.Data;
using System.Linq;
using System.Collections.Generic;
using System.Collections;
using System;
using System.Collections.Specialized;

namespace Ks.Common
{
    public delegate void NotifyCollectionChangedEventHandler<T>(object sender, NotifyCollectionChangedEventArgs<T> e);

    public class NotifyCollectionChangedEventArgs<T> : NotifyCollectionChangedEventArgs
    {

#pragma warning disable CS0618 // Type or member is obsolete
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action) : base(action)
        {
            this.InitializeAdd(action, null, -1);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateReset()
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Reset);
        }

        /// <summary>
        /// For Add, Remove, or Reset.
        /// </summary>
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, T changedItem) : base(action, changedItem)
        {
            if (action == NotifyCollectionChangedAction.Reset)
                this.InitializeAdd(action, null, -1);
            else
                this.InitializeAddOrRemove(action, new T[] { changedItem }, -1);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateAdd(T changedItem)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Add, changedItem);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateRemove(T changedItem)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Remove, changedItem);
        }

        /// <summary>
        /// For Add, Remove, or Reset.
        /// </summary>
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, T changedItem, int index) : base(action, changedItem, index)
        {
            if (action == NotifyCollectionChangedAction.Reset)
                this.InitializeAdd(action, null, -1);
            else
                this.InitializeAddOrRemove(action, new T[] { changedItem }, index);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateAdd(T changedItem, int index)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Add, changedItem, index);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateRemove(T changedItem, int index)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Remove, changedItem, index);
        }

        /// <summary>
        /// For Add, Remove, or Reset.
        /// </summary>
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, IList<T> changedItems) : base(action, (IList) changedItems)
        {
            if (action == NotifyCollectionChangedAction.Reset)
                this.InitializeAdd(action, null, -1);
            else
                this.InitializeAddOrRemove(action, changedItems, -1);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateAdd(IList<T> changedItems)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Add, changedItems);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateRemove(IList<T> changedItems)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Remove, changedItems);
        }

        /// <summary>
        /// For Add, Remove, or Reset.
        /// </summary>
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, IList<T> changedItems, int startingIndex) : base(action, (IList) changedItems, startingIndex)
        {
            if (action == NotifyCollectionChangedAction.Reset)
                this.InitializeAdd(action, null, -1);
            else
                this.InitializeAddOrRemove(action, changedItems, startingIndex);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateAdd(IList<T> changedItems, int startingIndex)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Add, changedItems, startingIndex);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateRemove(IList<T> changedItems, int startingIndex)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Remove, changedItems, startingIndex);
        }

        /// <summary>
        /// For Move or Replace.
        /// </summary>
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, T newItem, T oldItem) : base(action, newItem, oldItem)
        {
            this.InitializeMoveOrReplace(action, new T[] { newItem }, new T[] { oldItem }, -1, -1);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateReplace(T newItem, T oldItem)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Replace, newItem, oldItem);
        }

        /// <summary>
        /// For Move or Replace.
        /// </summary>
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, T newItem, T oldItem, int index) : base(action, newItem, oldItem, index)
        {
            this.InitializeMoveOrReplace(action, new T[] { newItem }, new T[] { oldItem }, index, index);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateReplace(T newItem, T oldItem, int index)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Replace, newItem, oldItem, index);
        }

        /// <summary>
        /// For Move or Replace.
        /// </summary>
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, IList<T> newItems, IList<T> oldItems) : base(action, (IList) newItems, (IList) oldItems)
        {
            this.InitializeMoveOrReplace(action, newItems, oldItems, -1, -1);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateMove(T changedItem)
        {
            var Items = new T[] { changedItem };
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Move, Items, Items);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateMove(IList<T> changedItems)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Move, changedItems, changedItems);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateReplace(IList<T> newItems, IList<T> oldItems)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Replace, newItems, oldItems);
        }

        /// <summary>
        /// For Move or Replace.
        /// </summary>
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, IList<T> newItems, IList<T> oldItems, int startingIndex) : base(action, (IList) newItems, (IList) oldItems, startingIndex)
        {
            this.InitializeMoveOrReplace(action, newItems, oldItems, startingIndex, startingIndex);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateReplace(IList<T> newItems, IList<T> oldItems, int startingIndex)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Replace, newItems, oldItems, startingIndex);
        }

        /// <summary>
        /// For Move or Replace.
        /// </summary>
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, T changedItem, int index, int oldIndex) : base(action, changedItem, index, oldIndex)
        {
            var changedItems = new T[] { changedItem };
            this.InitializeMoveOrReplace(action, changedItems, changedItems, index, oldIndex);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateMove(T changedItem, int index, int oldIndex)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Move, changedItem, index, oldIndex);
        }

        /// <summary>
        /// For Move or Replace.
        /// </summary>
        [Obsolete("Use the Create methods instead.")]
        public NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction action, IList<T> changedItems, int index, int oldIndex) : base(action, (IList) changedItems, index, oldIndex)
        {
            this.InitializeMoveOrReplace(action, changedItems, changedItems, index, oldIndex);
        }

        public static NotifyCollectionChangedEventArgs<T> CreateMove(IList<T> changedItems, int index, int oldIndex)
        {
            return new NotifyCollectionChangedEventArgs<T>(NotifyCollectionChangedAction.Move, changedItems, index, oldIndex);
        }
#pragma warning restore CS0618 // Type or member is obsolete

        public static NotifyCollectionChangedEventArgs<T> FromNotifyCollectionChangedEventArgs(NotifyCollectionChangedEventArgs E)
        {
            switch (E.Action)
            {
                case NotifyCollectionChangedAction.Move:
                    return CreateMove(E.NewItems.Cast<T>().ToArray(), E.NewStartingIndex, E.OldStartingIndex);
                case NotifyCollectionChangedAction.Replace:
                    return CreateReplace(E.NewItems.Cast<T>().ToArray(), E.OldItems.Cast<T>().ToArray(), E.NewStartingIndex);
                case NotifyCollectionChangedAction.Reset:
                    return CreateReset();
                case NotifyCollectionChangedAction.Add:
                    return CreateAdd(E.NewItems.Cast<T>().ToArray(), E.NewStartingIndex);
                case NotifyCollectionChangedAction.Remove:
                    return CreateRemove(E.OldItems.Cast<T>().ToArray(), E.OldStartingIndex);
            }

            Assert.Fail();
            return null;
        }

        public static NotifyCollectionChangedEventArgs<T> FromNotifyCollectionChangedEventArgsNoIndex(NotifyCollectionChangedEventArgs E)
        {
            switch (E.Action)
            {
                case NotifyCollectionChangedAction.Move:
                    return CreateMove(E.NewItems.Cast<T>().ToArray());
                case NotifyCollectionChangedAction.Replace:
                    return CreateReplace(E.NewItems.Cast<T>().ToArray(), E.OldItems.Cast<T>().ToArray());
                case NotifyCollectionChangedAction.Reset:
                    return CreateReset();
                case NotifyCollectionChangedAction.Add:
                    return CreateAdd(E.NewItems.Cast<T>().ToArray());
                case NotifyCollectionChangedAction.Remove:
                    return CreateRemove(E.OldItems.Cast<T>().ToArray());
            }

            Assert.Fail();
            return null;
        }

        private void InitializeAddOrRemove(NotifyCollectionChangedAction action, IList<T> changedItems, int startingIndex)
        {
            if (action == NotifyCollectionChangedAction.Add)
                this.InitializeAdd(action, changedItems, startingIndex);
            else if (action == NotifyCollectionChangedAction.Remove)
                this.InitializeRemove(action, changedItems, startingIndex);
        }

        private void InitializeAdd(NotifyCollectionChangedAction action, IList<T> newItems, int newStartingIndex)
        {
            this.NewItems = (newItems == null) ? null : new List<T>(newItems).AsReadOnly();
        }

        private void InitializeRemove(NotifyCollectionChangedAction action, IList<T> oldItems, int oldStartingIndex)
        {
            this.OldItems = (oldItems == null) ? null : new List<T>(oldItems).AsReadOnly();
        }

        private void InitializeMoveOrReplace(NotifyCollectionChangedAction action, IList<T> newItems, IList<T> oldItems, int startingIndex, int oldStartingIndex)
        {
            this.InitializeAdd(action, newItems, startingIndex);
            this.InitializeRemove(action, oldItems, oldStartingIndex);
        }

        public IList<T> ItemsGotIn
        {
            get
            {
                if ((this.Action == NotifyCollectionChangedAction.Add) | (this.Action == NotifyCollectionChangedAction.Replace))
                    return this.NewItems;
                return Utilities.Typed<T>.EmptyArray;
            }
        }

        public IList<T> ItemsWentOut
        {
            get
            {
                if ((this.Action == NotifyCollectionChangedAction.Remove) | (this.Action == NotifyCollectionChangedAction.Replace))
                    return this.OldItems;
                return Utilities.Typed<T>.EmptyArray;
            }
        }

        public new IList<T> NewItems { get; private set; }

        public new IList<T> OldItems { get; private set; }
    }

    public interface INotifyCollectionChanged<T> : INotifyCollectionChanged
    {
        new event NotifyCollectionChangedEventHandler<T> CollectionChanged;
    }

    public interface INotifyingCollection<T> : INotifyCollectionChanged<T>, ICollection<T>
    {
    }

    public class NotifyingList<T> : BaseList<T>, INotifyingCollection<T>
    {
        public NotifyingList(IList<T> List)
        {
            this.BaseList = List;
        }

        public NotifyingList() : this(new List<T>())
        {
        }

        public override void Clear()
        {
            this.BaseList.Clear();
            this.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateReset());
        }

        public override void Insert(int index, T item)
        {
            this.BaseList.Insert(index, item);
            this.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateAdd(item, index));
        }

        public override void RemoveAt(int index)
        {
            var OldItem = this.BaseList[index];
            this.BaseList.RemoveAt(index);
            this.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateRemove(OldItem, index));
        }

        public virtual void Move(int OldIndex, int NewIndex)
        {
            var Item = this.BaseList[OldIndex];
            this.BaseList.Move(OldIndex, NewIndex);
            this.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateMove(Item, NewIndex, OldIndex));
        }

        public virtual void SetFrom(IEnumerable<T> Collection)
        {
            this.BaseList.Clear();
            this.BaseList.AddRange(Collection);
            this.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateReset());
        }

        protected override IEnumerator<T> _GetEnumerator()
        {
            return this.BaseList.GetEnumerator();
        }

        public IEnumerator<T> GetEnumerator()
        {
            return this.BaseList.GetEnumerator();
        }

        public override int Count
        {
            get
            {
                return this.BaseList.Count;
            }
        }

        public override T this[int index]
        {
            get
            {
                return this.BaseList[index];
            }
            set
            {
                var OldItem = this.BaseList[index];
                this.BaseList[index] = value;
                this.OnCollectionChanged(NotifyCollectionChangedEventArgs<T>.CreateReplace(value, OldItem, index));
            }
        }

        public event NotifyCollectionChangedEventHandler<T> CollectionChanged;
        private event NotifyCollectionChangedEventHandler INotifyCollectionChanged_CollectionChanged;
        event NotifyCollectionChangedEventHandler INotifyCollectionChanged.CollectionChanged
        {
            add => this.INotifyCollectionChanged_CollectionChanged += value;
            remove => this.INotifyCollectionChanged_CollectionChanged -= value;
        }

        protected virtual void OnCollectionChanged(NotifyCollectionChangedEventArgs<T> E)
        {
            INotifyCollectionChanged_CollectionChanged?.Invoke(this, E);
            CollectionChanged?.Invoke(this, E);
        }

        private readonly IList<T> BaseList;
    }
}
